### PR TITLE
Move decoding to separate thread and generalize

### PR DESCRIPTION
**Pull request by jacobmcnamee**

### CODE REVIEW

**Comment by jacobmcnamee** [comment]
I need to go through and add some comments, but functionally everything should be good.

File: src/track.c [hunk 1]
```diff
@@ -58,6 +59,13 @@ char lock_detect_params_string[24] = LD_PARAMS_DISABLE;
 bool use_alias_detection = true;
 
 #define CN0_EST_LPF_CUTOFF 5
+#define GPS_WEEK_LENGTH_ms (7*24*60*60*1000)
```
Comment (ljbade) [inline src/track.c]: Can you use `WEEK_SECS` from `libswiftnav/time.h`?

e.g. `#define GPS_WEEK_LENGTH_ms (WEEK_SECS*1000)`

File: src/track.h [hunk 2]
```diff
@@ -71,6 +95,8 @@ typedef struct {
   s8 elevation;                /**< Elevation angle, degrees */
   alias_detect_t alias_detect; /**< Alias lock detector. */
   lock_detect_t lock_detect;   /**< Phase-lock detector state. */
+  nav_bit_fifo_t nav_bit_fifo;
```
Comment (ljbade) [inline src/track.h]: Add doxygen.

File: src/track.h [hunk 3]
```diff
@@ -71,6 +95,8 @@ typedef struct {
...
+  nav_time_sync_t nav_time_sync;
```
Comment (ljbade) [inline src/track.h]: Add doxygen.

File: src/track.c [hunk 4]
```diff
@@ -658,4 +799,22 @@ void tracking_setup()
   SETTING("track", "alias_detect", use_alias_detection, TYPE_BOOL);
 }
 
+bool tracking_channel_nav_bit_get(u8 channel, s8 *soft_bit)
+{
+  tracking_channel_t* chan = &tracking_channel[channel];
```
Comment (ljbade) [inline src/track.c]: * on wrong side of space

File: src/track.c [hunk 5]
```diff
@@ -658,4 +799,22 @@ void tracking_setup()
...
+}
+
+bool tracking_channel_time_sync(u8 channel, s32 TOW_ms, s8 bit_polarity)
+{
+  assert(TOW_ms >= 0);
+  assert(TOW_ms < GPS_WEEK_LENGTH_ms);
+  assert((bit_polarity == BIT_POLARITY_NORMAL) ||
+         (bit_polarity == BIT_POLARITY_INVERTED));
+
+  tracking_channel_t* chan = &tracking_channel[channel];
```
Comment (ljbade) [inline src/track.c]: * on wrong side of space

**Comment by ljbade** [comment]
Aside from the in-line comments ^, it looks good to me

**Comment by scottswift** [comment]
are we ready to hitl test this?

**Comment by jacobmcnamee** [comment]
Addresses issues, added a bunch of comments, a couple slight code changes.

**Comment by jacobmcnamee** [comment]
retest this please

**Comment by benjamin0** [comment]
retest this please

**Comment by benjamin0** [comment]
retest this please

**Comment by benjamin0** [comment]
retest this please

**Comment by benjamin0** [comment]
retest this please

**Comment by benjamin0** [comment]
retest this please

**Comment by jacobmcnamee** [comment]
Ready for testing. Any comments on the updated docs are welcome.

**Comment by jacobmcnamee** [comment]
Ready for testing. Any comments on the updated docs are welcomed.

**Comment by ljbade** [comment]
Everything looks great.

**Comment by fnoble** [comment]
We had a conversation on slack about how this would break ephemeris sending as they would all get sent at once - one idea to fix this is to not send ephemerides when they are received but just have a separate thread that transmits the current ephemeris at a (configurable?) interval. No reason it has to be done at time of reception.

**Comment by jacobmcnamee** [comment]
retest this please

**Comment by jacobmcnamee** [comment]
We did a quick field test confirm that the ephemeris messages were acceptably spaced as to not disrupt the radio link. Everything looked good.

**Comment by ljbade** [comment]
Cool I can finish my PRs

**Pull request merged**